# Pandas Advanced

Topics Covered:
1. Filtering Data
2. Sorting Data
3. Multiple Conditions
4. GroupBy

## 1. Filtering Data

Filtering means selecting only the rows that match a condition.

Example:
- Students with marks greater than 80
- Students with marks less than 70
- Students whose marks are exactly 90

Why it is useful:
In real datasets we often need to find specific records instead of looking at the entire dataset.

In [1]:
import pandas as pd

data = {
    "Name": ["Amit", "Riya", "John", "Sara", "Neha"],
    "Marks": [85, 72, 90, 65, 88]
}

df = pd.DataFrame(data)

df

,Name,Marks
0,Amit,85
1,Riya,72
2,John,90
3,Sara,65
4,Neha,88


### Example 1: Students scoring more than 80 marks

Condition:
Marks > 80

In [2]:
df[df["Marks"] > 80]

,Name,Marks
0,Amit,85
2,John,90
4,Neha,88


Explanation:

Only rows where Marks are greater than 80 are displayed.

85 → Included
72 → Not Included
90 → Included
65 → Not Included
88 → Included

In [3]:
df[df["Marks"] < 80]

,Name,Marks
1,Riya,72
3,Sara,65


In [4]:
df[df["Marks"] == 90]

,Name,Marks
2,John,90


Key Concept:

General Syntax:

df[condition]

Examples:

Marks greater than 80:
df[df["Marks"] > 80]

Marks less than 80:
df[df["Marks"] < 80]

Marks equal to 90:
df[df["Marks"] == 90]

Comparison Operators:

>   Greater than

<   Less than

>=  Greater than or equal to

<=  Less than or equal to

==  Equal to

!=  Not equal to

## 2. Sorting Data

Sorting means arranging data in ascending or descending order.

Examples:
- Lowest marks to highest marks
- Highest marks to lowest marks

Why it is useful:
Sorting helps us quickly identify top and bottom performers.

### Example 1: Sort Marks in Ascending Order

Ascending Order = Smallest to Largest

In [5]:
df.sort_values("Marks")

,Name,Marks
3,Sara,65
1,Riya,72
0,Amit,85
4,Neha,88
2,John,90


Explanation:

The marks are arranged from lowest to highest.

65 → Sara
72 → Riya
85 → Amit
88 → Neha
90 → John

### Example 2: Sort Marks in Descending Order

Descending Order = Largest to Smallest

In [6]:
df.sort_values("Marks", ascending=False)

,Name,Marks
2,John,90
4,Neha,88
0,Amit,85
1,Riya,72
3,Sara,65


Explanation:

The marks are arranged from highest to lowest.

90 → John
88 → Neha
85 → Amit
72 → Riya
65 → Sara

Key Concept:

df.sort_values("Column_Name")

Ascending:
df.sort_values("Marks")

Descending:
df.sort_values("Marks", ascending=False)

## 3. Multiple Conditions

Sometimes we need to apply more than one condition while filtering data.

Operators:

&  → AND (both conditions must be true)

|  → OR (at least one condition must be true)

Why it is useful:
Real-world datasets often require filtering based on multiple criteria.

### Example 1: Marks greater than 70 AND less than 90

Condition:

Marks > 70 AND Marks < 90

In [7]:
df[(df["Marks"] > 70) & (df["Marks"] < 90)]

,Name,Marks
0,Amit,85
1,Riya,72
4,Neha,88


Explanation:

Only students whose marks are greater than 70 and less than 90 are displayed.

85 → Included
72 → Included
90 → Not Included
65 → Not Included
88 → Included

### Example 2: Marks less than 70 OR equal to 90

Condition:

Marks < 70 OR Marks == 90

In [8]:
df[(df["Marks"] < 70) | (df["Marks"] == 90)]

,Name,Marks
2,John,90
3,Sara,65


Explanation:

At least one condition must be true.

65 → Included
90 → Included

Key Concept:

AND:

df[(condition1) & (condition2)]

OR:

df[(condition1) | (condition2)]

## 4. GroupBy()

GroupBy is used to group data based on a column and perform calculations on each group.

Examples:
- Average marks by department
- Total sales by city
- Number of employees in each department

Why it is useful:
It helps summarize and analyze large datasets.

### Create a New Dataset for GroupBy

In [9]:
data = {
    "Department": ["AI", "AI", "CS", "CS", "AI"],
    "Marks": [85, 90, 75, 80, 88]
}

df_dept = pd.DataFrame(data)

df_dept

,Department,Marks
0,AI,85
1,AI,90
2,CS,75
3,CS,80
4,AI,88


### Example 1: Calculate Average Marks by Department

In [10]:
df_dept.groupby("Department")["Marks"].mean()

Department
AI    87.666667
CS    77.500000
Name: Marks, dtype: float64

Explanation:

AI Department:
(85 + 90 + 88) / 3 = 87.67

CS Department:
(75 + 80) / 2 = 77.5

GroupBy creates groups and mean() calculates the average for each group.

### Example 2: Calculate Maximum Marks by Department

In [11]:
df_dept.groupby("Department")["Marks"].max()

Department
AI    90
CS    80
Name: Marks, dtype: int64

Explanation:

AI → 90

CS → 80

max() returns the highest value from each group.

Key Concept:

Group data:
df.groupby("Column")

Average:
df.groupby("Column")["Marks"].mean()

Maximum:
df.groupby("Column")["Marks"].max()

## Summary

In this section, we learned:

✅ Grouping data

✅ Calculating averages using mean()

✅ Finding maximum values using max()

✅ Basic data aggregation using GroupBy